modules/authtoservices.cpp

```diff
@@ -0,0 +1,491 @@
+            PutModule(t_s("We will connect even if SASL fails"));
+        }
+    }
+
+    void TrySaslCommand(const CString& sLine) {
+        if(!sLine.Token(1).empty()) {
+            SetNV("TrySASL", sLine.Token(1));
+        }
+
+        if(TrySasl()) {
+            PutModule(t_s("We attempt SASL negotiation"));
+        } else {
+            PutModule(t_s("We will not attempt SASL negotiation"));
+        }
+    }
+
+    void SaslMechanismCommand(const CString& sLine) {
+        CString sMechanisms = sLine.Token(1, true).AsUpper();
+
+        if (!sMechanisms.empty()) {
+            VCString vsMechanisms;
+            sMechanisms.Split(" ", vsMechanisms);
+
+            for (const CString& sMechanism : vsMechanisms) {
+                if (!SupportsMechanism(sMechanism)) {
+                    PutModule("Unsupported mechanism: " + sMechanism);
+                    return;
+                }
+            }
+
+            SetNV("SASLMechanisms", sMechanisms);
+        }
+
+        PutModule(t_f("Current mechanisms set: {1}")(GetMechanismsString()));
+    }
+
+    bool SupportsMechanism(const CString& sMechanism) const {
+        for (const auto& it : SupportedMechanisms) {
+            if (sMechanism.Equals(it.szName)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    CString GetMechanismsString() const {
+        if (GetNV("SASLMechanisms").empty()) {
+            CString sDefaults = "";
+
+            for (const auto& it : SupportedMechanisms) {
+                if (it.bDefault) {
+                    if (!sDefaults.empty()) {
+                        sDefaults += " ";
+                    }
+
+                    sDefaults += it.szName;
+                }
+            }
+
+            return sDefaults;
+        }
+
+        return GetNV("SASLMechanisms");
+    }
+
+    void PrintHelp(const CString& sLine) {
+        HandleHelpCommand(sLine);
+
+        CTable Mechanisms;
+        Mechanisms.AddColumn(t_s("Mechanism"));
+        Mechanisms.AddColumn(t_s("Description"));
+
+        for (const auto& it : SupportedMechanisms) {
+            Mechanisms.AddRow();
+            Mechanisms.SetCell(t_s("Mechanism"), it.szName);
+            Mechanisms.SetCell(t_s("Description"), it.sDescription.Resolve());
+        }
+
+        PutModule(t_s("The following mechanisms are available:"));
+        PutModule(Mechanisms);
+    }
+
+    ~CAuthToServices() override {}
+
+    bool LoadOtherModuleNV(const CString &otherModule, MCString& result) {
+        CString sPath = GetNetwork()->GetNetworkPath() + "/moddata/"
+                        + otherModule + "/.registry";
+        return result.ReadFromDisk(sPath) == MCString::MCS_SUCCESS;
+    }
+
+    bool OnLoad(const CString& sArgs, CString& sMessage) override {
+        if (!sArgs.empty() && sArgs != "<hidden>") {
+            SetNV("Password", sArgs);
+            SetArgs("<hidden>");
+        }
+
+        if(!GetNV("NotFirstLoad").ToBool()) {
```

Comment:
This module should be automatically loaded if sasl and/or nickserv was loaded before. That can be done in IRCNetwork.cpp (search for "legacy crap"). Also the migrated settings can mimic the previous settings more closely if you take into account which of those modules were loaded.

Also that could help getting rid of "NotFirstLoad" hack.